### ISSUE TITLE
Running Ansible separately from Vagrant

### ISSUE THREAD

**Issue by danbohea** [hashbangcode/vlad#202]
In fixing #192 today I noticed that running Ansible separately from Vagrant now requires some [extra bits & bobs](http://vlad-docs.readthedocs.org/en/dev/usage/ansible/). E.g:

```
ansible-playbook -i vlad/host.ini vlad/playbooks/site.yml --extra-vars "ansible_ssh_user=vagrant ansible_ssh_private_key_file=~/.vagrant.d/insecure_private_key"
```
Yuck. Not the most friendly one-liner ever but certainly better than having 2 thirds of our triggered playbooks not running.

This issue is for discussing & tracking having our cake & eating it: all triggered plays run as expected and `ansible-playbook` commands return to a sane length.

For the record, when we originally moved `hostfile = ./vlad/host.ini` and `inventory = ./vlad/host.ini` into ansible.cfg it never really worked - I've always had to re-include the `-i` option to get `ansible-playbook` to work. This latest development seems to be a further extension of that.

Why does running `ansible-playbook` not pick up the defaults that we're setting elsewhere?

**Comment by philipnorton42** [comment]
The main problem is that the configuration file isn't picked up all the time, it depends on what directory you are in at the time. From the ansible docs:

```
* ANSIBLE_CONFIG (an environment variable)
* ansible.cfg (in the current directory)
* .ansible.cfg (in the home directory)
* /etc/ansible/ansible.cfg
```
I think we might be able to put the configuration in the /vlad directory, but this would have to be duplicated both inside and outside of this directory.

The Ansible documentation on Vagrant provisioning states that the following should be used, which is even worse!

    ansible-playbook -i .vagrant/provisioners/ansible/inventory/vagrant_ansible_inventory --private-key=.vagrant/machines/default/virtualbox/private_key -u vagrant playbook.yml

**Comment by danbohea** [comment]
LOL, yeah I saw that particular one-liner in my own research - at least we're not having to wrestle with that monster :)

Duplicating the config file to both places does seem a little odd but I think it would be worth it to help simplify these commands. Let's aim to get this done & tested in time for v1.1.

**Comment by philipnorton42** [comment]
Yup, I'm currently running some tests on the centos builds but I'll add the new config file in as well.

**Comment by danbohea** [comment]
Jubbly.

**Comment by philipnorton42** [comment]
I've just pushed a script that should go towards solving this issue. The script called vlad-run.sh (name not set in stone) can be run to run the unusable playbook for the box. It's located in the scripts directory of the vlad folder and can be run from anywhere within the Vlad folders, For example, if you are in the docroot directory:

    ../vlad/scripts/vlad-run.sh

Additionally, I've added a single parameter to the script that allows you to run Ansible tags within Vlad. So, to run the Apache tag you would do something like this:

    ../vlad/scripts/vlad-run.sh apache

This script also doesn't rely on the hosts.ini inventory file, instead using the vagrant inventory file.

This is an idea I've been pondering for a while now but the use case here is clear :) What do you think?

**Comment by danbohea** [comment]
I'm not sure I'm keen on the idea of hiding (wrapping) Ansible. I think this is partly to do with me playing with Ansible for a fair while now, enjoying it and being glad it wasn't hidden from me.

That said, my view is singular and other folks might not give a monkeys and may prefer to see Ansible wrapped up and Vlad's non-Vagrant provisioning be a bit simpler.

I guess there's no harm in having both? Is there? Hmmm, I'm really torn on this.

**Comment by philipnorton42** [comment]
I can see where you are coming from. My take on this was just to provide an alternative to having to track down a bunch of flags that you had to copy and paste into the command line to get things to work. Also, in order to Ansible to pick up the correct configuration file it must be ran from a specific location, without this I think you would need to litter vlad with config files that changed depending on where they were (e.g., so that it always pointed to the same ansible.log location). Creating a simple script to wrap these issues seemed like a good idea. 

At least it means that I won't need to copy and paste a paragraph of text just to run an unable tag :)

This script shouldn't get any more complex than it is though. I think wrapping this single utility is more than enough.

**Comment by danbohea** [comment]
Sure. I guess the other bonus to wrapping things in a shell script like this is that it opens up the possibility of creating a simple bash alias (e.g. `vlad-play`) that expands to `vlad/scripts/vlad-run.sh` (or wherever you prefer to locate vlad/ in the context of you project - *(hot topic!)*).

I'm up for this approach so long as the docs still cover both approaches - that would address all concerns here.

**Comment by philipnorton42** [comment]
Cool. I'll update the docs with my bit, and I'll make sure that the rest of the Ansible docs make sense :)

**Comment by danbohea** [comment]
In light of that, might it make more sense to call the script `vlad-play.sh`? To me that makes it a teeny-weeny bit clearer as to what it might do (I can see some folks possibly incorrectly assuming that it might run Vagrant commands).

**Comment by philipnorton42** [comment]
Or even vlad-playbook.sh?

**Comment by danbohea** [comment]
LOL. At this rate we'll have added enough characters to make the original `ansible-playbook` command the shorter choice ;)

**Comment by danbohea** [comment]
Also, I'm guessing that this is a *nix-host-only luxury as it stands?

**Comment by philipnorton42** [comment]
Fair point :smiley: 
vlad-play.sh it is :P

**Comment by philipnorton42** [comment]
Yeah, we could always create a .bat version.

**Comment by danbohea** [comment]
Or put the bash script on the guest along with Ansible for Windows hosts? We could even create the alias I mentioned earlier.

**Comment by philipnorton42** [comment]
I have created that helper script and I've updated the documentation to show how to run things with it.

I was about to close this ben then I realised that we haven't get removed the hosts.ini file. Are we ok leaving this to after 1.1 is released?

**Comment by danbohea** [comment]
Sorry - don't follow you.

**Comment by philipnorton42** [comment]
We talked last week about the possibility of removing the host.ini file as the functionality we were using this for isn't really needed anymore. This file is actually generated by Vagrant itself so we are really replication functionality. It's just that removing the file would touch a few different areas of the project and we would need to alter the documentation as well.

**Comment by danbohea** [comment]
Let's the leave host.ini alone for now (pre 1.1) and then have another chat about it? I've only just looked at how you're getting inventory in the new vlad_play.sh script so I'm a little behind here :)

**Comment by philipnorton42** [comment]
Cool. Milestone changed :)

**Comment by danbohea** [comment]
There's a new dimension to this now that we're processing settings files with Ruby in order to provide Ansible with a single consistent filename/location (vlad_guts/merged_user_settings.yml).

As of current master (Vlad 1.1.4) if you were to make a change to your settings and then re-run a Vlad Ansible play separately from Vagrant, the change wouldn't get picked up (because the Vagrantfile now does this processing and Ansible would still be pointing at the unchanged merged_user_settings.yml). That may sound mad but this was all changed for good reasons, correcting things that were equally mad if not more mad.

Since we're already wrapping Ansible-playbook commands in a script for convenience, there's no reason why we couldn't repeat this settings file processing there too. Perhaps we could even abstract the code that processes the settings files so that this script and the Vagrant file can share it.
